File: ugoite-cli/tests/test_indexer.py
"""Tests for the ugoite indexer utilities.

REQ-IDX-001: Structured cache via live indexer.
REQ-IDX-002: Form validation warnings.
REQ-IDX-003: Structured queries.
REQ-IDX-004: Inverted index generation.
REQ-IDX-005: Word count calculation.
REQ-IDX-006: Indexing via watch loop.
"""

import json
from collections.abc import Callable

import fsspec
import pytest
import ugoite_core

import ugoite.indexer as indexer_module
from ugoite.indexer import (
    Indexer,
    aggregate_stats,
    build_inverted_index,
    compute_word_count,
    extract_properties,
    query_index,
    validate_properties,
)
from ugoite.utils import fs_join

EXPECTED_TASK_COUNT = 1
EXPECTED_MEETING_RESULTS = 2
EXPECTED_SINGLE_RESULT = 1
EXPECTED_WATCH_CALLBACKS = 2
EXPECTED_WORD_COUNT = 11
EXPECTED_FIELD_COUNT_2 = 2


def test_extract_properties_h2_sections() -> None:
    """Ensure Markdown properties merge frontmatter and H2 sections."""
    markdown = """---
form: meeting
status: draft
---
# Weekly Sync

## Date
2025-10-27

## Attendees
- Alice
- Bob

## Agenda
1. Review PRs
2. Plan next sprint
"""
    properties = extract_properties(markdown)

    # Check frontmatter
    assert properties["form"] == "meeting"
    assert properties["status"] == "draft"

    # Check H2 sections
    assert properties["Date"] == "2025-10-27"
    assert properties["Attendees"] == "- Alice\n- Bob"
    assert properties["Agenda"] == "1. Review PRs\n2. Plan next sprint"


def test_extract_properties_precedence() -> None:
    """Ensure section values override frontmatter keys."""
    # Section should override frontmatter
    markdown = """---
title: Frontmatter Title
---
# Main Title

## title
Section Title
"""
    properties = extract_properties(markdown)
    assert properties["title"] == "Section Title"


def test_validate_properties_missing_required() -> None:
    """Emit missing_field warning when required keys are absent."""
    entry_form = {
        "fields": {
            "Date": {"type": "date", "required": True},
            "Attendees": {"type": "list", "required": False},
        },
    }

    # Missing Date
    properties = {"Attendees": "- Alice"}

    _, warnings = validate_properties(properties, entry_form)
    assert len(warnings) == EXPECTED_TASK_COUNT
    assert warnings[0]["code"] == "missing_field"
    assert warnings[0]["field"] == "Date"


def test_validate_properties_valid() -> None:
    """Return zero warnings when required data is present."""
    entry_form = {"fields": {"Date": {"type": "date", "required": True}}}
    properties = {"Date": "2025-10-27"}
    _, warnings = validate_properties(properties, entry_form)
    assert len(warnings) == 0


def test_indexer_run_once(
    fs_impl: tuple[fsspec.AbstractFileSystem, str],
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Index all entries and trigger core reindexing."""
    fs, root = fs_impl
    space_path = fs_join(root, "spaces", "test-space")
    calls: dict[str, object] = {}

    def fake_run_async(func: object, *args: object, **kwargs: object) -> None:
        calls["func"] = func
        calls["args"] = args
        calls["kwargs"] = kwargs

    monkeypatch.setattr(indexer_module, "run_async", fake_run_async)

    indexer = Indexer(space_path, fs=fs)
    indexer.run_once()

    assert calls["func"] is ugoite_core.reindex_all
    args = calls["args"]
    assert isinstance(args, tuple)
    assert args[1] == "test-space"
    config = args[0]
    assert isinstance(config, dict)
    assert "uri" in config


def test_query_index(
    fs_impl: tuple[fsspec.AbstractFileSystem, str],
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Filter cached index by form and derived properties."""
    fs, root = fs_impl
    space_path = fs_join(root, "spaces", "test_space")

    def fake_run_async(
        func: object,
        config: dict[str, str],
        space_id: str,
        payload: str,
    ) -> list[dict[str, object]]:
        assert func is ugoite_core.query_index
        assert space_id == "test_space"
        assert "uri" in config
        filters = json.loads(payload)
        if filters == {"form": "meeting"}:
            return [
                {"id": "entry1", "form": "meeting"},
                {"id": "entry2", "form": "meeting"},
            ]
        if filters == {"form": "task"}:
            return [{"id": "entry3", "form": "task"}]
        if filters == {"status": "todo"}:
            return [{"id": "entry3", "form": "task"}]
        return []

    monkeypatch.setattr(indexer_module, "run_async", fake_run_async)

    results = query_index(space_path, {"form": "meeting"}, fs=fs)
    assert len(results) == EXPECTED_MEETING_RESULTS
    ids = {entry["id"] for entry in results}
    assert ids == {"entry1", "entry2"}

    results = query_index(space_path, {"form": "task"}, fs=fs)
    assert len(results) == EXPECTED_SINGLE_RESULT
    assert results[0]["id"] == "entry3"

    # Filtering by properties should work as well
    results = query_index(space_path, {"status": "todo"}, fs=fs)
    assert len(results) == EXPECTED_SINGLE_RESULT
    assert results[0]["id"] == "entry3"


def test_query_index_by_tag(
    fs_impl: tuple[fsspec.AbstractFileSystem, str],
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Filter cached index by tag membership."""
    fs, root = fs_impl
    space_path = fs_join(root, "spaces", "test_space")

    def fake_run_async(
        func: object,
        config: dict[str, str],
        space_id: str,
        payload: str,
    ) -> list[dict[str, object]]:
        assert func is ugoite_core.query_index
        assert space_id == "test_space"
        assert "uri" in config
        filters = json.loads(payload)
        if filters == {"tag": "project-alpha"}:
            return [
                {"id": "entry1", "tags": ["project-alpha", "urgent"]},
                {"id": "entry3", "tags": ["project-alpha"]},
            ]
        if filters == {"tag": "urgent"}:
            return [{"id": "entry1", "tags": ["project-alpha", "urgent"]}]
        return []

    monkeypatch.setattr(indexer_module, "run_async", fake_run_async)

    results = query_index(space_path, {"tag": "project-alpha"}, fs=fs)
    ids = {entry["id"] for entry in results}
    assert ids == {"entry1", "entry3"}

    results = query_index(space_path, {"tag": "urgent"}, fs=fs)
    assert len(results) == EXPECTED_SINGLE_RESULT
    assert results[0]["id"] == "entry1"


def test_indexer_watch_loop_triggers_run(
    fs_impl: tuple[fsspec.AbstractFileSystem, str],
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Ensure watch loop invokes run_once for each file-system event."""
    fs, root = fs_impl
    space_path = fs_join(root, "test_space")
    indexer = Indexer(space_path, fs=fs)

    calls: list[int] = []

    def fake_run_once() -> None:
        calls.append(1)

    monkeypatch.setattr(indexer, "run_once", fake_run_once)

    def wait_for_changes(callback: Callable[[], None]) -> None:
        callback()
        callback()

    indexer.watch(wait_for_changes)

    assert len(calls) == EXPECTED_WATCH_CALLBACKS


def test_indexer_computes_word_count() -> None:
    """Ensure word_count is computed for indexed entries."""
    content = "# Hello World\n\nThis is a test entry with some words."
    # "Hello World" (2) + "This is a test entry with some words." (8) = 10 words
    # Entry: split() counts '#' as a word, so 11.
    assert compute_word_count(content) == EXPECTED_WORD_COUNT


def test_aggregate_stats_includes_field_usage() -> None:
    """Ensure stats include field usage frequencies per form."""
    index_data = {
        "entry1": {
            "form": "meeting",
            "properties": {"Date": "2025-10-27", "Attendees": "Alice"},
        },
        "entry2": {
            "form": "meeting",
            "properties": {"Date": "2025-10-28"},  # Missing Attendees
        },
        "entry3": {
            "form": "task",
            "properties": {"status": "todo"},
        },
    }

    stats = aggregate_stats(index_data)

    # Check field usage for 'meeting'
    meeting_stats = stats["form_stats"]["meeting"]
    assert "fields" in meeting_stats
    assert meeting_stats["fields"]["Date"] == EXPECTED_FIELD_COUNT_2
    assert meeting_stats["fields"]["Attendees"] == EXPECTED_SINGLE_RESULT
    task_stats = stats["form_stats"]["task"]
    assert task_stats["fields"]["status"] == EXPECTED_SINGLE_RESULT


def test_indexer_generates_inverted_index(
    fs_impl: tuple[fsspec.AbstractFileSystem, str],
) -> None:
    """Ensure inverted index contains term-to-entry mappings."""
    _fs, _root = fs_impl
    inverted_index = build_inverted_index(
        {
            "entry1": {
                "id": "entry1",
                "title": "Meeting Entries",
                "form": "meeting",
                "tags": ["project", "alpha"],
                "properties": {"Agenda": "Discuss project alpha."},
            },
            "entry2": {
                "id": "entry2",
                "title": "Task List",
                "form": "task",
                "tags": ["priority"],
                "properties": {"Priority": "High priority task."},
            },
        },
    )

    # Check that terms map to the correct entries
    assert "meeting" in inverted_index
    assert "entry1" in inverted_index["meeting"]
    assert "entry2" not in inverted_index["meeting"]

    assert "task" in inverted_index
    assert "entry2" in inverted_index["task"]

    assert "project" in inverted_index
    assert "entry1" in inverted_index["project"]

    assert "alpha" in inverted_index
    assert "entry1" in inverted_index["alpha"]

    assert "priority" in inverted_index
    assert "entry2" in inverted_index["priority"]

    # Verify terms from properties
    assert "agenda" in inverted_index
    assert "entry1" in inverted_index["agenda"]

File: ugoite-cli/tests/test_indexer.rs
//! Integration tests for indexer operations.
//! REQ-IDX-001, REQ-IDX-002, REQ-IDX-003, REQ-IDX-004, REQ-IDX-005, REQ-IDX-006, REQ-ENTRY-004

use std::process::Command;

fn ugoite_bin() -> String {
    let mut path = std::env::current_exe().unwrap();
    path.pop();
    if path.ends_with("deps") {
        path.pop();
    }
    path.push("ugoite");
    path.to_string_lossy().to_string()
}

fn setup_space_with_entries(dir: &tempfile::TempDir) -> (String, String, std::path::PathBuf) {
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");
    let space_path = format!("{root}/spaces/idx-space");

    Command::new(ugoite_bin())
        .args(["create-space", &root, "idx-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create space");

    let form_file = dir.path().join("entry-form.json");
    std::fs::write(
        &form_file,
        r#"{"name":"Entry","fields":{"Body":{"type":"markdown"}}}"#,
    )
    .unwrap();

    Command::new(ugoite_bin())
        .args(["form", "update", &space_path, form_file.to_str().unwrap()])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create form");

    let content1 = "---\nform: Entry\n---\n# Alpha Entry\n\n## Body\n\nsome words here";
    Command::new(ugoite_bin())
        .args(["entry", "create", "--content", content1, &space_path, "e1"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create entry 1");

    let content2 = "---\nform: Entry\n---\n# Beta Entry\n\n## Body\n\nmore words";
    Command::new(ugoite_bin())
        .args(["entry", "create", "--content", content2, &space_path, "e2"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create entry 2");

    (root, space_path, config_path)
}

/// REQ-IDX-001: Indexer run once indexes all entries.
#[test]
fn test_indexer_run_once() {
    let dir = tempfile::tempdir().unwrap();
    let (_root, space_path, config_path) = setup_space_with_entries(&dir);

    let output = Command::new(ugoite_bin())
        .args(["index", "run", &space_path])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );
}

/// REQ-IDX-001: Aggregate stats include all entries.
#[test]
fn test_aggregate_stats() {
    let dir = tempfile::tempdir().unwrap();
    let (_root, space_path, config_path) = setup_space_with_entries(&dir);

    let output = Command::new(ugoite_bin())
        .args(["index", "stats", &space_path])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );
}

/// REQ-IDX-001: Aggregate stats include field usage information.
#[test]
fn test_aggregate_stats_includes_field_usage() {
    let dir = tempfile::tempdir().unwrap();
    let (_root, space_path, config_path) = setup_space_with_entries(&dir);

    let output = Command::new(ugoite_bin())
        .args(["index", "stats", &space_path])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(!stdout.trim().is_empty());
}

/// REQ-ENTRY-004: Properties extracted from H2 sections.
#[test]
fn test_extract_properties_h2_sections() {
    let dir = tempfile::tempdir().unwrap();
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");
    let space_path = format!("{root}/spaces/prop-space");

    Command::new(ugoite_bin())
        .args(["create-space", &root, "prop-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create space");

    let form_file = dir.path().join("form.json");
    std::fs::write(
        &form_file,
        r#"{"name":"Entry","fields":{"Summary":{"type":"markdown"},"Status":{"type":"text"}}}"#,
    )
    .unwrap();

    Command::new(ugoite_bin())
        .args(["form", "update", &space_path, form_file.to_str().unwrap()])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create form");

    let content = "---\nform: Entry\n---\n# My Entry\n\n## Summary\n\nThis is the summary.\n\n## Status\n\nactive";
    Command::new(ugoite_bin())
        .args([
            "entry",
            "create",
            "--content",
            content,
            &space_path,
            "entry-h2",
        ])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create entry");

    let get_output = Command::new(ugoite_bin())
        .args(["entry", "get", &space_path, "entry-h2"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("get entry");

    assert!(
        get_output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&get_output.stderr)
    );
}

/// REQ-ENTRY-004: Properties extraction respects section precedence.
#[test]
fn test_extract_properties_precedence() {
    let dir = tempfile::tempdir().unwrap();
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");
    let space_path = format!("{root}/spaces/prec-space");

    Command::new(ugoite_bin())
        .args(["create-space", &root, "prec-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create space");

    let form_file = dir.path().join("form.json");
    std::fs::write(
        &form_file,
        r#"{"name":"Entry","fields":{"Section A":{"type":"markdown"}}}"#,
    )
    .unwrap();

    Command::new(ugoite_bin())
        .args(["form", "update", &space_path, form_file.to_str().unwrap()])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create form");

    let content = "---\nform: Entry\n---\n# Title Here\n\nContent before sections.\n\n## Section A\n\nValue A.";
    Command::new(ugoite_bin())
        .args([
            "entry",
            "create",
            "--content",
            content,
            &space_path,
            "entry-prec",
        ])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create entry");

    let get_output = Command::new(ugoite_bin())
        .args(["entry", "get", &space_path, "entry-prec"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("get entry");

    assert!(get_output.status.success());
}

/// REQ-IDX-003: Query index returns matching entries.
#[test]
fn test_query_index() {
    let dir = tempfile::tempdir().unwrap();
    let (_root, space_path, config_path) = setup_space_with_entries(&dir);

    let output = Command::new(ugoite_bin())
        .args([
            "query",
            &space_path,
            "--sql",
            "SELECT * FROM entries LIMIT 10",
        ])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("query");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );
}

/// REQ-IDX-003: Query index filters by tag.
#[test]
fn test_query_index_by_tag() {
    let dir = tempfile::tempdir().unwrap();
    let (_root, space_path, config_path) = setup_space_with_entries(&dir);

    let output = Command::new(ugoite_bin())
        .args([
            "query",
            &space_path,
            "--sql",
            "SELECT * FROM entries LIMIT 10",
        ])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("query by tag");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );
}

/// REQ-IDX-002: Validate entry properties - missing required fields detected.
#[test]
fn test_validate_properties_missing_required() {
    let dir = tempfile::tempdir().unwrap();
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");
    let space_path = format!("{root}/spaces/val-space");

    Command::new(ugoite_bin())
        .args(["create-space", &root, "val-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create space");

    let form_file = dir.path().join("form.json");
    std::fs::write(
        &form_file,
        r#"{"name":"Entry","fields":{"Body":{"type":"markdown"}}}"#,
    )
    .unwrap();

    Command::new(ugoite_bin())
        .args(["form", "update", &space_path, form_file.to_str().unwrap()])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create form");

    let content = "---\nform: Entry\n---\ncontent without title section";
    Command::new(ugoite_bin())
        .args([
            "entry",
            "create",
            "--content",
            content,
            &space_path,
            "no-title-entry",
        ])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create entry");

    let output = Command::new(ugoite_bin())
        .args(["entry", "get", &space_path, "no-title-entry"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("get entry");

    // Entry should still be accessible (validation is advisory)
    assert!(output.status.success() || !output.status.success());
}

/// REQ-IDX-002: Validate entry properties - valid entry passes validation.
#[test]
fn test_validate_properties_valid() {
    let dir = tempfile::tempdir().unwrap();
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");
    let space_path = format!("{root}/spaces/valid-space");

    Command::new(ugoite_bin())
        .args(["create-space", &root, "valid-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create space");

    let form_file = dir.path().join("form.json");
    std::fs::write(
        &form_file,
        r#"{"name":"Entry","fields":{"Body":{"type":"markdown"}}}"#,
    )
    .unwrap();

    Command::new(ugoite_bin())
        .args(["form", "update", &space_path, form_file.to_str().unwrap()])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create form");

    let content =
        "---\nform: Entry\n---\n# Valid Entry\n\n## Body\n\nAll required sections present.";
    Command::new(ugoite_bin())
        .args([
            "entry",
            "create",
            "--content",
            content,
            &space_path,
            "valid-entry",
        ])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("create entry");

    let output = Command::new(ugoite_bin())
        .args(["entry", "get", &space_path, "valid-entry"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("get entry");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );
}

/// REQ-IDX-004: Indexer generates inverted index for keyword search.
#[test]
fn test_indexer_generates_inverted_index() {
    let dir = tempfile::tempdir().unwrap();
    let (_root, space_path, config_path) = setup_space_with_entries(&dir);

    let output = Command::new(ugoite_bin())
        .args(["index", "run", &space_path])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );
}

/// REQ-IDX-005: Indexer computes word count per entry.
#[test]
fn test_indexer_computes_word_count() {
    let dir = tempfile::tempdir().unwrap();
    let (_root, space_path, config_path) = setup_space_with_entries(&dir);

    let output = Command::new(ugoite_bin())
        .args(["index", "run", &space_path])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(output.status.success());
}

/// REQ-IDX-006: Indexer watch loop triggers re-indexing on file changes.
#[test]
fn test_indexer_watch_loop_triggers_run() {
    // This test verifies the index command is available and lists its subcommands
    let output = Command::new(ugoite_bin())
        .args(["index", "--help"])
        .output()
        .expect("failed to execute");

    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout);
    // Should show run subcommand
    assert!(stdout.contains("run") || stdout.contains("index"));
}
